### ISSUE TITLE
Initial Discussion and TODOs for 1st release

### ISSUE THREAD

**Issue by nfischer** [shelljs/shx#1]
_*[NOTE: @nfischer opened this to discuss adding a README, and I'm totally hijacking it. Sorry - @ariporad]*_

Ok, this issue will serve as the replacement for shelljs/shelljs#244. Feel free to discuss or provide feedback.

Things that need to be done prior to 1st release (which will hopefully be in conjunction with `shelljs@0.7.0`):

- [x] Add README.md 
- [x] Add some tests (I already have a bunch on my local machine, I'll push to github in a sec. - @ariporad)
- [ ] Setup Travis (Need tests first)
- [x] Pick a license
- [x] Land shelljs/shelljs#357 (This will make testing a lot easier/better)
- [x] Land shelljs/shelljs#394
- [ ] ???

**Comment by nfischer** [comment]
Merging these is fine with me.

As for license, I'm fine with the MIT license (I license most of my stuff under that anyway). It might be confusing to have our stuff under two licenses (BSD for shelljs, MIT for shx), but I don't think that's a huge concern.

I can write up a basic README later today, that links back to shelljs and has simple examples

**Comment by ariporad** [comment]
@nfischer: I personally vote to change `shelljs` to MIT (with @arturadib's permission, of course). A good chunk of the node world is under the MIT, and you can't use BSD code in a MIT project, and I can't think of any reason we'd need the BSD over MIT.

**Comment by ariporad** [comment]
@nfischer: 
In response to your comment [here][old]):

> @ariporad the shjs bin file is still useful. It executes shelljs scripts without the need for modifying NODE_PATH, which is convenient for some (I know I found it convenient before I realized I could modify that).

 If I have a `scripts/my-script.js` file, aren't the following equivalent: `shjs scripts/my-script` and `node scripts/my-script`?

[old]: https://github.com/shelljs/shelljs/pull/244#issuecomment-184792422

**Comment by ariporad** [comment]
Ok, it looks like the tests are dependent on shelljs/shelljs#357.

**Comment by nfischer** [comment]
@ariporad the difference is if you have NODE_PATH set up (I think that's env car, but I may be mistaken). Try installing shelljs globally, unsetting NODE_PATH, and running a script from an arbitrary directory. "shjs" should always work, while "node" sometimes doesn't (it can't find the shelljs folder)

This was the case for me, because my global installs go to a folder other than /use/local

**Comment by nfischer** [comment]
@ariporad I just added a README. Feel free to modify it drastically, it's just an initial draft.

**Comment by levithomason** [comment]
Is there any interest in starting this project off with ES2015 using [babel](http://babeljs.io/) from the get go?

**Comment by ariporad** [comment]
@levithomason: I've been thinking about this. I've been considering switching shelljs to ES6, and if we did I'd probobly switch shx over at the same time.

I'm not really sure about switching shx over to babel first, it might be an interesting testing ground, and the tests are already es6. (I'm also trying to switch the shelljs tests to es6 and ava, so there's that).

Let me think on it.

**Comment by levithomason** [comment]
Awesome.  I'd love to help get things going on babel and a build system using just package scripts if it is of interest.  LMK

**Comment by nfischer** [comment]
@levithomason I've been contributing to dthree/cash a bit recently and really like their build system. Is something like that what you had in mind?

**Comment by levithomason** [comment]
Nice, I've done a lot of gulp build systems.  Lately, I've been picking up the trend of [using npm as a build tool](http://blog.keithcirkel.co.uk/how-to-use-npm-as-a-build-tool/) which [quite a few](https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=using+npm+as+a+build+tool) folks seem to be picking up on as well.

Initially, I had many qualms with this approach.  Mostly due to task orchestration difficulties.  After having tried it a few times now, I completely love it.  There are no extra deps to manage, and only one cli to work with, `npm`.

I still use gulp when the build system is complicated, like in [Stardust](https://github.com/TechnologyAdvice/stardust).  Though most the time a solid and fast build system can be done with just `npm` scripts, like [local-sync](https://github.com/TechnologyAdvice/local-sync).

For `shx` (and ShellJS) I think it'd be great to have ES6 code bases, Mocha or AVA tests, ESLint, and a vanilla `npm` build.  A workflow where the build and lint runs on watchers is great for speeding up development, too.

These are some of the things I've found delightful as of late.  Would love to hear what specifically you liked about dthree/cash.

**Comment by nfischer** [comment]
I really liked how their project uses watchers that automatically run the build steps. The only issue is that everything crashes if you press save for a file with syntax errors.

**Comment by levithomason** [comment]
Oh yea, yuck.  When using gulp with watchers and linters, I use [`gulp-plumber`](https://www.npmjs.com/package/gulp-plumber) to prevent the build from crashing but still log the error.

I'd love to scaffold something out if we're ready for it.

**Comment by nfischer** [comment]
@levithomason now is probably a good time. We're still getting things set up, but it'd be nice to have this project ready to go (initial release) by the time ShellJS `v0.7` is released.

**Comment by ariporad** [comment]
+1
On Sun, Mar 6, 2016 at 8:57 PM, Nate Fischer <notifications@github.com> wrote:
@levithomason [https://github.com/levithomason] now is probably a good time. We're still getting things set up, but it'd be
nice to have this project ready to go (initial release) by the time ShellJS v0.7 is released.

—
Reply to this email directly or view it on GitHub
[https://github.com/shelljs/shx/issues/1#issuecomment-193099922] .[https://github.com/notifications/beacon/ABu7pKuy3uJ9TkjJki5CkacUXKNQ5Xy7ks5pq7AwgaJpZM4Ha2ZD.gif]

**Comment by levithomason** [comment]
Looks like there are 15 open items for 0.7.  Can't imagine setting up the build would take any more than an evening, a few at most.  If you want to push your tests...

>- [ ] Add some tests (I already have a bunch on my local machine, I'll push to github in a sec. - @.ariporad)

...and give the OK, I can start a PR for the base build.  Probably split it up into a few PRs, one for each hunk.  Babel, lint, test, etc.

**Comment by nfischer** [comment]
@levithomason 
> Looks like there are 15 open items for 0.7.

I'm going to discuss with @ariporad about bumping some of these back to v0.8 (or whatever the next release is), so that we can get this released sooner.

**Comment by levithomason** [comment]
Sounds good, ping me if you'd like any help.

**Comment by nfischer** [comment]
If you can get a build system set up for this project, that should be more than helpful enough. Also, feel free to review any PRs @ariporad and I put out for shelljs :+1: Thanks for the help!

**Comment by levithomason** [comment]
Have a go at #4 and LMK what you think.  It is a minimal vanilla npm build system hitting the points we talked about.

**Comment by ariporad** [comment]
@levithomason: Thanks so much for your PR!

**Comment by nfischer** [comment]
Just added the MIT license (I see no reason not to go with that one)

**Comment by levithomason** [comment]
Agreed

**Comment by levithomason** [comment]
Looks like I don't have write perms to the repo, is this intended or oversight?  I've setup tests and was trying to open a PR here (closed my fork).

I can setup my fork again if needed, but would rather have write rights if at all possible.

**Comment by ariporad** [comment]
No, that was an oversight. Hang on a sec.

**Comment by ariporad** [comment]
@shelljs/shx: Better?

**Comment by levithomason** [comment]
Cool, will try that later.  Moved back to fork for now.  Also, I'll be traveling this week so not sure of my availability.  Two PRs open for now, one for tests and another for the shelljs github dep.

Catch you guys later, cheers!

**Comment by levithomason** [comment]
Yes, I have rights :D

**Comment by ariporad** [comment]
Awesome! No rush either.


Ari

**Comment by levithomason** [comment]
Most the items in this issue have been addressed.  The remaining were pulled into separate issues.  Closing.
